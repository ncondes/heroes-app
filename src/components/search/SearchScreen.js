import { useMemo } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string'
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';


export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse( location.search );

    const initialState = {
        searchText: q,
    }

    const [ values, handleInputChange ] = useForm( initialState );
    
    const { searchText } = values;

    const heroesFiltered = useMemo(() => getHeroesByName( q ), [ q ])

    const handleSubmit = (e) => {
        e.preventDefault();
        navigate(`?q=${ searchText }`)
    }

    return (
        <>
            <h1>Search</h1>
            <hr />

            <div className='row'>
                <div className='col-4 d-flex flex-column'>
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={ handleSubmit }>
                        <input
                            type='text'
                            placeholder='Search a hero'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value= { searchText }
                            onChange={ handleInputChange }
                        />
                        <div className='d-grid gap-2'>
                            <button
                                type='submit'
                                className='btn btn-outline-dark mt-3'
                            >
                                Search
                            </button>
                        </div>
                    </form>
                </div>

                <div className='col-8 d-flex flex-column'>
                    <h4>Results</h4>
                    <hr />

                    {
                        ( q === '') 
                            ? <div className='alert alert-dark'>Search a Hero in the form</div>
                            : ( heroesFiltered.length === 0 )
                                && <div className='alert alert-danger'>No results for: { q } </div>
                    }

                    {
                        heroesFiltered.map( heroe => (
                            <HeroCard
                                key={ heroe.id }
                                { ...heroe }
                            />
                        ))
                    }
                </div>
            </div>
        </>
    )
}
