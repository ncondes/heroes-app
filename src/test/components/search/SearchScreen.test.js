import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';


const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Pruebas en SearchScreen', () => {

    test('Debe de mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        )

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-dark').text().trim() ).toBe( 'Search a Hero in the form' )

    });

    test('Debe de mostrar a batman y el input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <SearchScreen />
            </MemoryRouter>
        )

        expect( wrapper.find('input').prop('value') ).toBe( 'batman' );
        expect( wrapper ).toMatchSnapshot();

    });
    
    test('Debe de mostrar un error si no se encuentra el hero', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman123'] }>
                <SearchScreen />
            </MemoryRouter>
        )

        expect( wrapper.find('.alert-danger').text().trim() ).toBe( 'No results for: batman123' )

    });
    
    test('Debe de llamar el navigate a la nueva pantalla', () => {
    
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman',
            },
        })

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        })

        expect( mockNavigate ).toHaveBeenCalledWith( '?q=batman' );

    })
    
})