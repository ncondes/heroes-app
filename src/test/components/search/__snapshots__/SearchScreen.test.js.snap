// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen Debe de mostrar a batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-4 d-flex flex-column"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-dark mt-3"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-8 d-flex flex-column"
    >
      <h4>
        Results
      </h4>
      <hr />
      <a
        className="my-card animate__animated animate__fadeIn"
        href="/hero/dc-batman"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive"
          src="./assets/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera aparici√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>,
]
`;

exports[`Pruebas en SearchScreen Debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-4 d-flex flex-column"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-dark mt-3"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-8 d-flex flex-column"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-dark"
      >
        Search a Hero in the form
      </div>
    </div>
  </div>,
]
`;
